// td-upyun/lib/create-req.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var rq = require("request-promise");
var Req = /** @class */ (function () {
    function Req(options) {
        this.baseURL = options.endpoint + "/" + options.service.serviceName;
        this.service = options.service;
        this.getHeaderSign = options.getHeaderSign;
    }
    Req.prototype.get = function (uri, options) {
        return rq.get(encodeURI("" + this.baseURL + uri), this.genOptions(uri, 'GET', options));
    };
    Req.prototype.post = function (uri, options) {
        return rq.post(encodeURI("" + this.baseURL + uri), this.genOptions(uri, 'POST', options));
    };
    Req.prototype.put = function (uri, options) {
        return rq.put(encodeURI("" + this.baseURL + uri), this.genOptions(uri, 'PUT', options));
    };
    Req.prototype.head = function (uri, options) {
        return rq.head(encodeURI("" + this.baseURL + uri), this.genOptions(uri, 'HEAD', options));
    };
    Req.prototype.delete = function (uri, options) {
        return rq.delete(encodeURI("" + this.baseURL + uri), this.genOptions(uri, 'DELETE', options));
    };
    Req.prototype.patch = function (uri, options) {
        return rq.patch(encodeURI("" + this.baseURL + uri), this.genOptions(uri, 'PATCH', options));
    };
    Req.prototype.genOptions = function (uri, method, options) {
        var headers = this.getHeaderSign(this.service, method, uri);
        return Object.assign({}, options, {
            resolveWithFullResponse: true,
            method: method,
            headers: headers,
            proxy: this.service.proxy
        });
    };
    return Req;
}());
exports.Req = Req;
function default_1(options) {
    return new Req(options);
}
exports.default = default_1;
