// td-upyun/lib/sign.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var md5 = require("md5");
var crypto = require("crypto");
var base64 = require("base-64");
var _version = '3.3.4';
function hmacsha1(key, data) {
    return crypto.createHmac('sha1', key).update(data).digest().toString('base64');
}
/**
 * generate head sign
 * @param {object} service
 * @param {string} path - storage path on upyun server, e.g: /your/dir/example.txt
 * @param {string} contentMd5 - md5 of the file that will be uploaded
 */
function getHeaderSign(service, method, path, contentMd5) {
    var date = new Date().toUTCString();
    path = '/' + service.serviceName + path;
    var sign = genSign(service, {
        method: method,
        path: path,
        date: date,
        contentMd5: contentMd5
    });
    return {
        'Authorization': sign,
        'X-Date': date,
    };
}
exports.getHeaderSign = getHeaderSign;
/**
 * generate signature string which can be used in head sign or body sign
 * @param {object} service
 * @param {object} options - must include key is method, path
 */
function genSign(service, options) {
    var method = options.method, path = options.path;
    var data = [
        method,
        encodeURI(path)
    ];
    // optional params
    ['date', 'policy', 'contentMd5'].forEach(function (item) {
        if (options[item]) {
            data.push(options[item]);
        }
    });
    // hmacsha1 return base64 encoded string
    var sign = hmacsha1(service.password, data.join('&'));
    return "UPYUN " + service.operatorName + ":" + sign;
}
exports.genSign = genSign;
/**
 * get policy and authorization for form api
 * @param {object} service
 * @param {object} - other optional params @see http://docs.upyun.com/api/form_api/#_2
 */
function getPolicyAndAuthorization(service, params) {
    params['service'] = service.serviceName;
    if (typeof params['save-key'] === 'undefined') {
        throw new Error('upyun - calclate body sign need save-key');
    }
    if (typeof params['expiration'] === 'undefined') {
        // default 30 minutes
        params['expiration'] = Math.floor(new Date().getTime() / 1000 + 30 * 60);
    }
    var policy = base64.encode(JSON.stringify(params));
    var authorization = genSign(service, {
        method: 'POST',
        path: '/' + service.serviceName,
        policy: policy
    });
    return {
        policy: policy,
        authorization: authorization
    };
}
exports.getPolicyAndAuthorization = getPolicyAndAuthorization;
function getPurgeHeaderSign(service, urls) {
    var date = new Date().toUTCString();
    var str = urls.join('\n');
    var sign = md5(str + "&" + service.serviceName + "&" + date + "&" + service.password);
    return {
        'Authorization': "UpYun " + service.serviceName + ":" + service.operatorName + ":" + sign,
        'Date': date,
        'User-Agent': 'Js-Sdk/' + _version
    };
}
exports.getPurgeHeaderSign = getPurgeHeaderSign;
